protocols/kad/src/behaviour.rs

```diff
@@ -2492,6 +2547,37 @@ pub enum KademliaEvent {
     PendingRoutablePeer { peer: PeerId, address: Multiaddr },
 }
 
+/// Information about progress events.
+#[derive(Debug, Clone)]
+pub struct ProgressStep {
+    /// The index into the event
+    pub count: NonZeroUsize,
+    /// Is this the final event?
+    pub last: bool,
+}
+
+impl ProgressStep {
+    fn first() -> Self {
+        Self {
+            count: NonZeroUsize::new(1).expect("1 to be greater than 0."),
+            last: false,
+        }
+    }
+
+    fn first_and_last() -> Self {
+        let mut first = ProgressStep::first();
+        first.last = true;
+        first
+    }
+
+    fn next(&self) -> Self {
+        assert!(!self.last);
+        let count = NonZeroUsize::new(self.count.get() + 1).expect("Not to exhaust usize space.");
```

Comment:
It's a bit more complicated, overflow checks are only inserted in `debug` builds, in `release` builds this will wrap around ( and the above program you posted will pass)